### PR TITLE
Add mozillaclub backend

### PR DESCRIPTION
**Pull request by acs**

### CODE REVIEW

**Comment by acs** [comment]
@sduenas ready for the final review!

File: perceval/backends/mozilla/mozillaclub.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+        return False
+
+    @staticmethod
+    def metadata_id(item):
+        """Extracts the identifier from an event item."""
+
+        try:
+            a = item.keys()
+        except AttributeError:
+            print ("->"+item+"<-")
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: Remove this line and I think the `try..catch` too

File: perceval/backends/mozilla/mozillaclub.py [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+        cells = sheet_json['feed']['entry']
+
+        event_fields = self.__get_event_fields(cells)
+
+        # Process all events reading the rows according to the event template
+        # The only way to detect the end of row is looking to the
+        # number of column. When the max number is reached (cell_cols) the next
+        # cell is from the next row.
+        # The first while is for looping rows and the second columns
+        while cells:
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: I don't like the whole idea of this loop. You're using `__get_next_event` to change the contents of `cells`. This is not a good practice because it is hard to follow what the code is doing. Methods should not change the status of an object that is given as parameter.

One possible solution is to create an Interable object which is initialized with the cells and which iteraters over cells, parsing them.

File: perceval/backends/mozilla/mozillaclub.py [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+#     Alvaro del Castillo <acs@bitergia.com>
+#
+
+import json
+import logging
+import os.path
+
+import requests
+
+
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: Remove this blank line

File: perceval/backends/mozilla/mozillaclub.py [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+from ...backend import Backend, BackendCommand, metadata
+from ...cache import Cache
+from ...errors import BackendError, CacheError, ParseError
+
+from ...utils import (DEFAULT_DATETIME,
+                     datetime_to_utc,
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: Bad indentation of these imports.

File: perceval/backends/mozilla/mozillaclub.py [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+    :param tag: label used to mark the data
+    """
+    version = '0.1.0'
+
+    def __init__(self, url=MOZILLA_CLUB_URL, cache=None, tag=None):
+        origin = url
+        self.url = url
+        super().__init__(origin, tag=tag, cache=cache)
+        self.client = MozillaClubClient(url)
+        self.__users = {}  # internal users cache
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: Do you need this object?

File: perceval/backends/mozilla/mozillaclub.py [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+
+        nevents = 0  # number of events processed
+        nevents_wrong = 0  # number of events with wrong data
+
+        self._purge_cache_queue()
+
+        raw_cells = self.client.get_cells()
+        self._push_cache_queue(raw_cells)
+        sheet_json = json.loads(raw_cells)
+        self._flush_cache_queue()
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: This statement should be executed once the parsed has returned all the events. Otherwise it would lead to a cache with invalid data when the parser fails.

File: perceval/backends/mozilla/mozillaclub.py [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+        This backend only generates one type of item which is
+        'event'.
+        """
+        return 'event'
+
+    @staticmethod
+    def metadata_updated_on(item):
+        """Extracts the update time from a MozillaClub item.
+
+        The timestamp is extracted from 'end' field.
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: This is not the field you are using to generate the result.

File: perceval/backends/mozilla/mozillaclub.py [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+
+        req = requests.get(uri)
+        req.raise_for_status()
+
+        return req.text
+
+    def get_cells(self):
+        """Retrieve all cells from the spreadsheet."""
+
+        logging.info("Retrieving all cells spreadsheet data ...")
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: Do not use `logging`, use `logger` object instead.

File: perceval/backends/mozilla/mozillaclub.py [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+        raw_cells = self.call(self.url)
+
+        return raw_cells
+
+
+class MozillaClubParser:
+    """Git log parser.
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: This is not a Git parser.

Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: I would like to have in the documentation some examples about how the format of the input is. It is not enough clear how this is parsed and since parsing is a little tricky, it would be useful to have some examples here. Please check Git parser to find good examples.

File: perceval/backends/mozilla/mozillaclub.py (hunk 3 above)
Comment (acs) [inline perceval/backends/mozilla/mozillaclub.py]: Done!

File: perceval/backends/mozilla/mozillaclub.py (hunk 5 above)
Comment (acs) [inline perceval/backends/mozilla/mozillaclub.py]: Not anymore. Removed.

File: perceval/backends/mozilla/mozillaclub.py (hunk 1 above)
Comment (acs) [inline perceval/backends/mozilla/mozillaclub.py]: Done!

File: perceval/backends/mozilla/mozillaclub.py (hunk 7 above)
Comment (acs) [inline perceval/backends/mozilla/mozillaclub.py]: Fixed. Here it is a problem: this date is not the real update date for the item. Probably we must use the unixtime for the collect timestamp. What do you think @sduenas?

File: perceval/backends/mozilla/mozillaclub.py (hunk 8 above)
Comment (acs) [inline perceval/backends/mozilla/mozillaclub.py]: Fixed!

File: perceval/backends/mozilla/mozillaclub.py (hunk 9 above)
Comment (acs) [inline perceval/backends/mozilla/mozillaclub.py]: Added a comment with the basic format of the list of cells and the content of a cell.

**Comment by acs** [comment]
@sduenas new version uploaded covering all comments! Could you review it? Thanks!

File: tests/test_mozillaclub.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+        mozillaclub = MozillaClub(MozillaClub_FEED_URL, cache=cache)
+
+        events = [event for event in mozillaclub.fetch()]
+        self.assertEqual(len(http_requests), 1)
+
+        # Now, we get the events from the cache.
+        # The contents should be the same and there won't be
+        # any new request to the server
+        cached_events = [event for event in mozillaclub.fetch_from_cache()]
+        self.assertEqual(len(cached_events), len(events))
```
Comment (sduenas) [inline tests/test_mozillaclub.py]: You should compare the contents of `cache_events` and `events`. Not just their length.

File: tests/test_mozillaclub.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        body = read_file('data/mozillaclub/feed.json')
+        httpretty.register_uri(httpretty.GET,
+                               MozillaClub_FEED_URL,
+                               body=body, status=200)
+
+        client = MozillaClubClient(MozillaClub_FEED_URL)
+        response = client.get_cells()
+
+        self.assertEqual(response, body)
+
```
Comment (sduenas) [inline tests/test_mozillaclub.py]: I miss here the tests for the parser class. All classes must have their own unit tests.

File: perceval/backends/mozilla/mozillaclub.py [hunk 12]
```diff
@@ -0,0 +1,441 @@
...
+        "title": {
+            "$t": "A1",
+            "type": "text"
+        },
+        "updated": {
+            "$t": "2016-12-13T15:44:04.821Z"
+        }
+    }
+    """
+
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: Remove this blank line.

File: perceval/backends/mozilla/mozillaclub.py [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+    def fetch(self):
+        """Fetch events from the MozillaClub url.
+
+        The method retrieves, from a MozillaClub url, the
+        events. The data is a Google spreadsheet retrieved using
+        the feed API REST.
+
+        :returns: a generator of events
+        """
+
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: Remove this blank line.

File: perceval/backends/mozilla/mozillaclub.py [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+        origin = url
+        self.url = url
+        super().__init__(origin, tag=tag, cache=cache)
+        self.client = MozillaClubClient(url)
+
+    @metadata
+    def fetch(self):
+        """Fetch events from the MozillaClub url.
+
+        The method retrieves, from a MozillaClub url, the
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: Replace _url_ by _URL_. Do the same in the next line.

File: perceval/backends/mozilla/mozillaclub.py [hunk 15]
```diff
@@ -0,0 +1,441 @@
...
+    @metadata
+    def fetch_from_cache(self):
+        """Fetch the events from the cache.
+
+        :returns: a generator of events
+
+        :raises CacheError: raised when an error occurs accessing the
+            cache
+        """
+
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: Remove this blank line.

File: perceval/backends/mozilla/mozillaclub.py [hunk 16]
```diff
@@ -0,0 +1,441 @@
...
+        :returns: this backend supports items resuming
+        """
+        return False
+
+    @staticmethod
+    def metadata_id(item):
+        """Extracts the identifier from an event item."""
+
+        return str(item['Date of Event']+"_"+item['Club Name'])
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: Let a space between operators.

File: perceval/backends/mozilla/mozillaclub.py [hunk 17]
```diff
@@ -0,0 +1,441 @@
...
+    """MozillaClub API client.
+
+    This class implements a simple client to retrieve events from
+    projects in a MozillaClub site.
+
+    :param url: URL of MozillaClub
+
+    :raises HTTPError: when an error occurs doing the request
+    """
+
```
Comment (sduenas) [inline perceval/backends/mozilla/mozillaclub.py]: Remove this blank line.

File: perceval/backends/mozilla/mozillaclub.py (hunk 16 above)
Comment (acs) [inline perceval/backends/mozilla/mozillaclub.py]: Fixed!

File: tests/test_mozillaclub.py (hunk 10 above)
Comment (acs) [inline tests/test_mozillaclub.py]: Yes, I will fix it asap.

File: tests/test_mozillaclub.py (hunk 11 above)
Comment (acs) [inline tests/test_mozillaclub.py]: Yes, I will fix it asap.

File: perceval/backends/mozilla/mozillaclub.py (hunk 14 above)
Comment (acs) [inline perceval/backends/mozilla/mozillaclub.py]: Done!

File: perceval/backends/mozilla/mozillaclub.py (hunk 13 above)
Comment (acs) [inline perceval/backends/mozilla/mozillaclub.py]: Done!

File: perceval/backends/mozilla/mozillaclub.py (hunk 15 above)
Comment (acs) [inline perceval/backends/mozilla/mozillaclub.py]: Done!

File: perceval/backends/mozilla/mozillaclub.py (hunk 17 above)
Comment (acs) [inline perceval/backends/mozilla/mozillaclub.py]: Done!

File: perceval/backends/mozilla/mozillaclub.py (hunk 12 above)
Comment (acs) [inline perceval/backends/mozilla/mozillaclub.py]: Done!

File: tests/test_mozillaclub.py (hunk 10 above)
Comment (acs) [inline tests/test_mozillaclub.py]: After reviewing it I compare the full contents in next line with:

    for i in range(0,len(events)):
            self.assertDictEqual(cached_events[i]['data'], events[i]['data'])

Maybe I am missing something ...

Comment (acs) [inline tests/test_mozillaclub.py]: Done!

File: tests/test_mozillaclub.py (hunk 11 above)
Comment (acs) [inline tests/test_mozillaclub.py]: Done!

**Comment by acs** [comment]
@sduenas all ready for another review round.

**Comment by acs** [comment]
Hmm, I think it is already added to README.md ... 

https://github.com/grimoirelab/perceval-mozilla/pull/2/files#diff-04c6e90faac2675aa89e2176d2eec7d8R42

Am I missing something? Is that in the general perceval README.md file?

**Comment by sduenas** [comment]
@acs there is another section named **Backends** on the top of the README file.

**Comment by sduenas** [comment]
Merged!
